File: QA_Interview_Questions/13-lru-cache.ts
/**
 * Question 13: Implement LRU Cache
 * Problem: Implement a Least Recently Used (LRU) Cache
 * 
 * Operations:
 * - get(key): Return value if key exists, else -1
 * - put(key, value): Update or insert key-value pair, evict LRU item if full
 */

class LRUCache {
  private maxSize: number;
  private cache: Map<number, number>;
  
  constructor(capacity: number) {
    this.maxSize = capacity;
    this.cache = new Map();
  }
  
  get(key: number): number {
    if (!this.cache.has(key)) {
      return -1;
    }
    
    // Move to end (most recently used)
    const value = this.cache.get(key)!;
    this.cache.delete(key);
    this.cache.set(key, value);
    
    return value;
  }
  
  put(key: number, value: number): void {
    // If key exists, remove it (will be re-added at end)
    if (this.cache.has(key)) {
      this.cache.delete(key);
    }
    
    // Add or update key-value pair at the end
    this.cache.set(key, value);
    
    // If cache exceeds max size, remove least recently used (first item)
    if (this.cache.size > this.maxSize) {
      const firstKey = this.cache.keys().next().value as number;
      this.cache.delete(firstKey);
    }
  }
}

// Test cases
const lruCache = new LRUCache(2);

lruCache.put(1, 1); // cache: {1=1}
lruCache.put(2, 2); // cache: {1=1, 2=2}
console.log(lruCache.get(1)); // 1

lruCache.put(3, 3); // cache: {1=1, 3=3}, evict 2
console.log(lruCache.get(2)); // -1

lruCache.put(4, 4); // cache: {3=3, 4=4}, evict 1
console.log(lruCache.get(1)); // -1
console.log(lruCache.get(3)); // 3
console.log(lruCache.get(4)); // 4

File: Java_Interview_Questions/13-LRUCache.java
/**
 * Question 13: Implement LRU Cache
 * Problem: Implement a Least Recently Used (LRU) Cache
 * 
 * Operations:
 * - get(key): Return value if key exists, else -1
 * - put(key, value): Update or insert, evict LRU if full
 */

import java.util.*;

public class LRUCache {
    
    private int maxSize;
    private Map<Integer, Integer> cache;
    
    public LRUCache(int capacity) {
        this.maxSize = capacity;
        // LinkedHashMap maintains insertion order
        this.cache = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry eldest) {
                return size() > maxSize;
            }
        };
    }
    
    public int get(int key) {
        return cache.getOrDefault(key, -1);
    }
    
    public void put(int key, int value) {
        cache.put(key, value);
    }
    
    @Override
    public String toString() {
        return cache.toString();
    }
}

// Manual implementation with Doubly Linked List
class LRUCacheManual {
    
    private class Node {
        int key;
        int value;
        Node prev;
        Node next;
        
        Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }
    
    private int maxSize;
    private Map<Integer, Node> cache;
    private Node head;
    private Node tail;
    
    public LRUCacheManual(int capacity) {
        this.maxSize = capacity;
        this.cache = new HashMap<>();
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        
        Node node = cache.get(key);
        removeNode(node);
        addToHead(node);
        return node.value;
    }
    
    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            Node node = cache.get(key);
            node.value = value;
            removeNode(node);
            addToHead(node);
        } else {
            Node newNode = new Node(key, value);
            cache.put(key, newNode);
            addToHead(newNode);
            
            if (cache.size() > maxSize) {
                Node lru = tail.prev;
                removeNode(lru);
                cache.remove(lru.key);
            }
        }
    }
    
    private void addToHead(Node node) {
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }
    
    private void removeNode(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
}

class LRUCacheTest {
    public static void main(String[] args) {
        // Test LinkedHashMap based implementation
        LRUCache lru = new LRUCache(2);
        
        lru.put(1, 1);
        lru.put(2, 2);
        System.out.println(lru.get(1)); // 1
        
        lru.put(3, 3); // Evict 2
        System.out.println(lru.get(2)); // -1
        
        lru.put(4, 4); // Evict 1
        System.out.println(lru.get(1)); // -1
        System.out.println(lru.get(3)); // 3
        System.out.println(lru.get(4)); // 4
        
        System.out.println("\n--- Manual Implementation ---");
        
        LRUCacheManual lruManual = new LRUCacheManual(2);
        
        lruManual.put(1, 1);
        lruManual.put(2, 2);
        System.out.println(lruManual.get(1)); // 1
        
        lruManual.put(3, 3); // Evict 2
        System.out.println(lruManual.get(2)); // -1
        
        lruManual.put(4, 4); // Evict 1
        System.out.println(lruManual.get(1)); // -1
        System.out.println(lruManual.get(3)); // 3
        System.out.println(lruManual.get(4)); // 4
    }
}
